python/ray/tests/test_autoscaler_yaml.py

```diff
@@ -134,6 +134,43 @@ def testInvalidConfig(self):
         self._test_invalid_config(
             os.path.join("tests", "additional_property.yaml"))
 
+    def testValidateCustomSecurityGroupConfig(self):
+        aws_config_path = os.path.join(RAY_PATH,
+                                       "autoscaler/aws/example-minimal.yaml")
+        with open(aws_config_path) as f:
+            config = yaml.safe_load(f)
+
+        # Test validate security group with custom permissions
+        ip_permissions = [{
+            "FromPort": port,
+            "ToPort": port,
+            "IpProtocol": "TCP",
+            "IpRanges": [{
+                "CidrIp": "0.0.0.0/0"
+            }],
+        } for port in [80, 443, 8265]]
+        config["provider"].update({
+            "security_group": {
+                "IpPermissions": ip_permissions
+            }
+        })
+        config = prepare_config(config)
+        try:
+            validate_config(config)
```

Comment:
can we assert that the names are what we think they are? also you may want to pass in `prepare_config(copy.deepcopy(config))` if you reuse it in the same test since it's pretty much undefined whether `prepare_config` will modify the argument in place.

Reply:
Updated as you suggested. Also just to make sure we are on the same page, `prepare_config` doesn't fill in security group names (or IpPermissions), it is generated here https://github.com/ray-project/ray/blob/eb7f5b71e07f78b7ef16e954527f0e59be1a1687/python/ray/autoscaler/_private/aws/config.py#L526.

Reply:
yeah my bad, this isn't on the prepare_config path. I'm a little confused then, how is this testing your code if we never call `bootstrap_config`?

Reply:
This doesn't test the logic of my code. I thought your request is about adding the schema and testing that. I'd need to test `_get_or_create_vpc_security_groups`, but I couldn't find unit tests for functions in `python/ray/autoscaler/_private/aws/config.py`. Can you point me to them?

Reply:
Here's an explicit one: https://github.com/ray-project/ray/blob/5e0231d4f025e9dc97dc7ca6c78882a1b4f0c396/python/ray/tests/aws/test_autoscaler_aws.py#L60

Also, our `get_or_create_head_node` tests provide some coverage of the code path (could be better, but it does end up catching bugs).

Reply:
I do see why you were originally thinking of putting this in `test_autoscaler_aws` now though...

Reply:
Yeah I added a test to `test_autoscaler_aws`